src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs

```diff
@@ -28,25 +31,134 @@ public override async Task<int> ExecuteAsync()
             try
             {
                 EnsureOptionsCompatibility(_options);
+                // use a set to accumulate dependencies as we go
+                HashSet<StrippedDependency> accumulatedDependencies = new HashSet<StrippedDependency>();
+                // at the end of each depth level, these are added to the queue to clone
+                Queue<StrippedDependency> dependenciesToClone = new Queue<StrippedDependency>();
+                // use a set to keep track of whether we've seen dependencies before, otherwise we get trapped in circular dependencies
+                HashSet<StrippedDependency> seenDependencies = new HashSet<StrippedDependency>();
+                // list of master copies of repos that we relocate the .gitdirs for later
+                HashSet<string> masterCopies = new HashSet<string>();
                 RemoteFactory remoteFactory = new RemoteFactory(_options);
 
                 if (string.IsNullOrWhiteSpace(_options.RepoUri))
                 {
                     Local local = new Local(Logger);
                     IEnumerable<DependencyDetail>  rootDependencies = await local.GetDependenciesAsync();
-
-                    foreach (DependencyDetail dep in rootDependencies)
-                    {
-                        Logger.LogInformation($"Found local dependency {dep.RepoUri}@{dep.Commit}.  Starting deep clone...");
-                        await CloneRemoteRepoAndDependencies(remoteFactory, _options.ReposFolder, dep.RepoUri, dep.Commit, _options.IncludeToolset, Logger);
-                    }
+                    IEnumerable<StrippedDependency> stripped = rootDependencies.Select(d => new StrippedDependency(d));
+                    stripped.ForEach((s) => accumulatedDependencies.Add(s));
+                    stripped.ForEach((s) => seenDependencies.Add(s));
```

Comment:
NIT: this is enumerating twice. Add to both lists in the same ForEach.

Reply:
Fixed.